Skills are selected on a per-job basis

afterthought2 opened the pull request:
And many back-end changes to the skill system. This is something I've been playing around with, inspired by [this](https://baystation12.net/forums/threads/real-skill-system.6217/) forum post. I think the only hope of having a skill system with in-game effects is to have the skill availability and selection tied to your job, so here is my attempt to implement that.

User-facing changes:

- Skills are chosen on a per-job basis by going to the "Occupations" tab and clicking on a job. This pulls up a screen where you allocate points. Some jobs have minimum skill levels, and the number of points varies by job.
- Some changes to the skills and their categories. There is an extra skill level now, for a total of five.
- Players will lose skill data currently saved!
- AI/borgs can't choose skills. They are currently assigned a default value. This would later be changed so the skill levels are determined by module selection for borgs.

Other changes:

- All mobs have skills, found in the mob/skillset variable, which is a datum with some getters for finding the values. Most mobs start with a generic default skill value for all possible skills.
- At least most transformations lead to inheriting skillsets from the old mob to the new one. I would require help to determine where else this needs work and what would make sense there.

If this was merged, I would start following up with PRs which use skill checks to influence game outcomes. The idea is that (for most skills) at skill level 0, performing some actions leads to dangerous failure outcomes, at level 1 to frequent and inconvenient failure outcomes, and at 2/3/4 to smaller debuffs/nothing/buffs.

I would welcome any advice for how to improve the code. Alternatively, spam that downvote button to voice your disapproval of the entire undertaking.

![skills](https://user-images.githubusercontent.com/36152617/38062668-4334f3fa-32c3-11e8-85ed-1021c68cf69b.png)

afterthought2 committed:
backend skill system changes

Job mins and skill descriptions

UI improvement

Species and age stuff, cleanup.

hierarchies working

Fixes conflicts. Reduces whitespace.

on-station antags handled

code improvements

more code improvements

conflicts fixed

color changes

EcklesFire commented:
Is there a maximum amount of skill points with this change?

afterthought2 commented:
Skill point allocations can be set on a per-job basis. The default is 10, but most jobs will have less (because they also give some free skill values which can't be removed, see the red on the CO in the screenshot). The numbers are pretty arbitrary at this point, and easy to change. See "Skill points remaining" in the screenshot, at the top.

You can also set a max amount of skills on a per-job per-skill basis, for example banning the CO from selecting "Professional" or "Experienced" in Anatomy.

CakeQ commented:
Good stuff

EcklesFire commented:
Would it be possible to make the amount of points vary based on age and species? The reasoning being that some species are far more knowledgeable/trained in comparison to their human equivalents and  also because of the inherent experience increase with age.

I would personally also like to see basic/trained tiers costing less skill point wise than lower tiers. The reason being that for much of the skillsets, the  reason being that some skills are intermingled/interchangeable and infer ability in one or the other to some extent. Also  that some skills just simply do not possess the same time investment for basic/decent skills that others do.

E.G. The Chief Engineer is going to be trained in leadership, possess basic bureaucracy skills,  EVA training, and they're going to be at least trained in each engineering category.

I'm actually not really digging anything that's particularly restrictive because a character can possess, well, quite a varied skillset but only be specialised in a single thing. I'd like to see that kept as a 'thing.' I'm not sure we'd be in a great spot if suddenly someone forgot how to do their taxes because they got too good at flying a shuttle, for instance.

afterthought2 commented:
@EcklesFire, my intention was that in-game, you would see a very large effect in going 0->1, a large effect going from 1->2, and then much smaller effects after that (this is not implemented here, of course). If you look at the minimum skill allocations, they are almost all for level 1 or 2, with the vast majority at 1. Jobs like engineer that require lots of skills are just given extra skill points, so they can decide whether to specialize or take points outside the main category or whatever.

The age/species thing shouldn't be hard to add, and if that's what people want I'll do it. The varying cost thing is also not hard. I was trying to keep the allocation process simple to avoid confusing everyone, though.

Jovaniph commented:
This is a start... The implementation I approve, but i'm sure we can tweak the skills around here and there later. +1

EcklesFire commented:
@afterthought2 

Fair enough for the skill bit part.

I would very much like to see the age/species point variance added, though, and I believe it is quite an important part of it in order for accurate skill representation.

mustafakalash commented on code/_helpers/global_lists.dm:
```diff
@@ -130,6 +134,18 @@ var/global/list/string_slot_flags = list(
 
+	//Skills
+	//list of skills, with entries /datum/skill categories and values lists of /datum/skill skill entries of that category
+	for(var/T in (typesof(/datum/skill)-/datum/skill))
```
`for(var/T in subtypesof(/datum/skill))`

afterthought2 replied:
While it sounds stupid, apparently `subtypesof()` doesn't work here (get a compilation error: unknown proc).

PsiOmegaDelta replied:
That's because the macro is defined after this line in code, so it doesn't exist yet.

mustafakalash commented on code/_helpers/global_lists.dm:
```diff
@@ -130,6 +134,18 @@ var/global/list/string_slot_flags = list(
+	for(var/T in (typesof(/datum/skill)-/datum/skill))
+		if (length(typesof(T)) > 1) //This means it's a nonempty category
+			var/datum/skill/S = new T
+			GLOB.skills_by_category += S
```
This isn't necessary. If the attribute doesn't exist when you set it, it will be added automatically.

mustafakalash commented on code/_helpers/global_lists.dm:
```diff
@@ -130,6 +134,18 @@ var/global/list/string_slot_flags = list(
+			var/datum/skill/S = new T
+			GLOB.skills_by_category += S
+			GLOB.skills_by_category[S] = list() //finding all the children for the value
+			for(var/child_path in (typesof(T)-T))
```
`subtypesof(T)` again

mustafakalash commented on code/_helpers/global_lists.dm:
```diff
@@ -130,6 +134,18 @@ var/global/list/string_slot_flags = list(
+	//Skills
+	//list of skills, with entries /datum/skill categories and values lists of /datum/skill skill entries of that category
+	for(var/T in (typesof(/datum/skill)-/datum/skill))
+		if (length(typesof(T)) > 1) //This means it's a nonempty category
```
Doesn't matter if it's empty, the loop just wouldn't run, and returning an empty list for an empty category would be more expected than a null for an existing category.

afterthought2 replied:
While it may be irrelevant that the category is nonempty, checking whether it's empty is the only way of knowing whether it's a category or a skill. Without it, skills are added to the list of categories, defeating the purpose.

mustafakalash commented on code/game/jobs/job_controller.dm:
```diff
@@ -58,6 +63,25 @@ var/global/datum/controller/occupations/job_master
+			return 0
+		for(var/datum/job/J in occupations)
+			job_min_skills += J
+			job_max_skills += J
```
These two lines aren't necessary.

mustafakalash commented on code/game/jobs/job_controller.dm:
```diff
@@ -357,11 +381,16 @@ var/global/datum/controller/occupations/job_master
 		if(job)
 
+			// Transfers the skill settings for the job to the mob
+			if (H.client.prefs.skills_allocated[job.title])
```
Might need to make sure client exists.

afterthought2 replied:
Didn't see this until now. This was actually just completely wrong. Forgot to change it at some point; fixed now. Whether there should be a client check early on in this whole proc is also a valid question, but note that there is an H.client proc call ten lines down anyway that's always been there.

mustafakalash commented on code/modules/client/preference_setup/occupation/occupation.dm:
```diff
@@ -27,6 +31,7 @@
 	S["player_alt_titles"] >> pref.player_alt_titles
 	S["char_branch"]       >> pref.char_branch
 	S["char_rank"]         >> pref.char_rank
+	S["skills_allocated"]  >> pref.skills_allocated
```
`from_file(S["skills_allocated"], pref.skills_allocated)`

mustafakalash commented on code/modules/client/preference_setup/occupation/occupation.dm:
```diff
@@ -36,10 +41,12 @@
 	S["player_alt_titles"] << pref.player_alt_titles
 	S["char_branch"]       << pref.char_branch
 	S["char_rank"]         << pref.char_rank
+	S["skills_allocated"]  << pref.skills_allocated
```
`to_file(S["skills_allocated"], pref.skills_allocated)`

mustafakalash commented on code/modules/client/preference_setup/occupation/occupation.dm:
```diff
@@ -242,6 +254,35 @@
+		var/datum/skill/S = locate(href_list["skillinfo"])
+		if(!istype(S))
+			return
+		var/HTML = "<h2>[S.name]</h2>"
```
`var/list/HTML = list("<h2>[S.name]</h2>)"` then use `jointext(HTML, null)` in the browse.

mustafakalash commented on code/modules/client/preference_setup/occupation/occupation.dm:
```diff
@@ -242,6 +254,35 @@
+		for(i=0, i <= SKILL_MAX, i++)
+			var/level_name = S.level_names[i+1]
+			HTML +=	"<br><b>[level_name]</b>: [S.desc_levels[level_name]]<br>"
+		user << browse(HTML, "window=\ref[user]skillinfo")
```
`show_browser(user, jointext(HTML, null), "...")`

mustafakalash commented on code/modules/client/preference_setup/occupation/skill_selection.dm:
```diff
@@ -0,0 +1,103 @@
+			var/cap   = min(max[S], level + points_by_job[job]) //if selecting the skill would make you overspend, it won't be shown
+			dat += "<tr style='text-align:left;'>"
+			dat += "<th><a href='?src=\ref[src];skillinfo=\ref[S]'>[S.name]</a></th>"
+			var/i
```
You can initialize the index in the for loop.

mustafakalash commented on code/modules/mob/skills/skill.dm:
```diff
@@ -0,0 +1,300 @@
+/datum/skill/medical
+	field = "Medical"
+
+var/global/list/SKILLS = null
```
This appears to be unused.

mustafakalash commented on code/modules/mob/skills/skillset.dm:
```diff
@@ -0,0 +1,71 @@
+		to_chat(user, "There are no skills to display.")
+		return
+
+	var/HTML = "<body>"
```
Use a list

mustafakalash commented on code/modules/mob/skills/skillset.dm:
```diff
@@ -0,0 +1,71 @@
+	HTML += "</table>"
+
+	user << browse(null, "window=preferences")
+	user << browse(HTML, "window=show_skills;size=700x900")
```
Use `show_browser`

mustafakalash reviewed (changes requested):
I like this change. I don't like the thought of priming for skills affecting the game, though. Anyway, here's a first pass.

afterthought2 commented:
Age now gives a buff to skill point allocation. Currently gives between -1 and 2 skill points, but can be configured on a per-species basis (currently defaulting to the human settings). Can be used to give some species a flat point buff/debuff as well.

Species can be configured to give buffs/debuffs to skill point allocations on a per-job basis. Currently no species are actually doing that. I suggest both of these are tuned later, once the skills start becoming more meaningful, in consultation with the relevant lore people.

Changing your age/species may reset your skills (if you lost points).

Also, various code cleanup stuff (thanks, mkalash, I should have addressed most of your comments). 
EDIT: Now with a fancier data structure, as suggested by Psi and Techhead.

PsiOmegaDelta commented on code/_helpers/global_lists.dm:
```diff
@@ -130,6 +134,17 @@ var/global/list/string_slot_flags = list(
+			for(var/child_path in (typesof(T)-T))
+				var/datum/skill/child = new child_path
+				GLOB.skills_by_category[S] += child
+				GLOB.skills += child			//Also initialize this at the same time
```
This seems very similar to what /decl/hierarchy does

PsiOmegaDelta commented on code/modules/client/preference_setup/occupation/skill_selection.dm:
```diff
@@ -0,0 +1,107 @@
+					sum += spent
+
+		points_by_job[job] = job.skill_points							//We compute how many points we had.
+		if(job.title != "AI" && job.title != "Cyborg")					//These guys don't get species/age buffs.
```
Please make this a var on the job, not a hardcoded check.

PsiOmegaDelta commented on code/modules/mob/mob.dm:
```diff
@@ -44,6 +44,7 @@
 
 /mob/Initialize()
 	. = ..()
+	skillset = new(src)
```
Must be qdel-ed in `/Destroy()`

PsiOmegaDelta commented on code/modules/mob/skills/skillset.dm:
```diff
@@ -0,0 +1,71 @@
+	return from_datum()
+
+/datum/skillset/proc/obtain_from_mob(mob/mob)
+	if(!istype(mob) || istype(mob, /mob/living/silicon) || istype(owner, /mob/living/silicon))	//We don't transfer skills to/from silicon mobs.
```
Type checks like this should be avoided. Have a proc/var that's overridden appropriately instead.

Alex6511 commented:
This looks cool but I don't like the idea of random RNG failures if you're bad at something. I don't really thing that RNG has a place inside SS13 or Bay.

That being said I like everything that's in this PR.

Jovaniph commented:
Why is the Chief Engineer getting less points than the Senior Engineer? That doesn't make sense.

afterthought2 commented:
Keep in mind that CE gets two extra free skills, so the total points are the same. The general philosophy was that the command positions get some points into skills like leadership and bureaucracy and then a similar setup to the rank and file, while the senior positions, like Senior Engineer and Researcher, just get a couple of extra points to spend however. The senior engineer might well be a better engineer than the CE, they just aren't a manager.

That being said, the specific values are pretty arbitrary and I would be happy to adjust them. I have been assuming that they would be tuned in future PRs, as skills become more relevant.

afterthought2 commented:
Conflicts fixed. The setup window is reverted back to pre-#20743 width, with formatting modified appropriately. Antags should be handled correctly now, giving default (Experienced) skill level in all skills.

PsiOmegaDelta commented on code/modules/mob/skills/skillset.dm:
```diff
@@ -0,0 +1,82 @@
+	var/skills_transferable = TRUE
+
+/datum/skillset/New(mob/mob)
+	if(istype(mob))
```
When would an owner ever not be a mob?

PsiOmegaDelta commented on code/modules/mob/skills/skill.dm:
```diff
@@ -0,0 +1,316 @@
+
+/decl/hierarchy/skill/organizational
+	name = "Organizational"
+	ID	 = "1"
```
Have to say I'm not a fan of the ID-matching you've got going on. In general you should get decls by their types and no other means. There is only one exception and that is when loading or saving character preferences as proper types cannot be saved.

However, that is a situation where text2path() might actually be valid, so long as appropriate sanity checks are made.

afterthought2 replied:
See if you like this better. Almost everything is using either decls or their paths now. The IO stuff is restricted to a pair of procs that use strings of the paths to store the save data, rather than ID/job title. The conversion is using string matching, but I could do text2path if you think it would be superior. I thought it's better to loop over the known list of decls rather than the unknown list from the savefile.
EDIT: and now those procs are called only at reasonable times, on save and load. Dependence on job_master is reduced, at least to the point where sanitization will run without it. The UI stuff won't open until later in the init, and checks for job_master there (this is legacy behavior, not something new).

PsiOmegaDelta commented on maps/torch/job/jobs.dm:
```diff
@@ -58,6 +58,12 @@
 	allowed_ranks = list(
 		/datum/mil_rank/ec/o6
 	)
+	min_skill = list(	"management" 	= SKILL_ADEPT,
```
And these mappings should definitely use type paths, not strings.

PsiOmegaDelta commented on code/modules/mob/skills/skillset.dm:
```diff
@@ -0,0 +1,82 @@
+
+	show_browser(user, null, "window=preferences")
+	show_browser(user, jointext(HTML, null), "window=show_skills;size=700x900")
+	return
```
Trailing return.

PsiOmegaDelta commented on code/modules/mob/mob_defines.dm:
```diff
@@ -161,4 +161,6 @@
\ No newline at end of file
+	var/nabbing = 0  // Whether a creature with a CAN_NAB tag is grabbing normally or in nab mode.
+	
+	var/datum/skillset/skillset
```
If you have this be `var/datum/skillset/skillset = datum/skillset` then borgs could have a specialized `/datum/skillset/silicon/` skillset subtype
```
/mob/living/silicon
    skillset = /datum/skillset/silicon
```
and change the creation proc to:
```
/mob/proc/create_skillset()
	skillset = new skillset(src)
```

Then you could have any and all modifications in them and throw out things like
```
/mob/living/silicon/create_skillset()
	. = ..()
	skillset.skills_transferable = FALSE
```

afterthought2 replied:
Done. This should help with borg module stuff in the future.

PsiOmegaDelta commented on code/modules/mob/skills/skill.dm:
```diff
@@ -0,0 +1,316 @@
+				GLOB.skills += C.children
+		else
+			log_error("<span class='warning'>Warning: multiple instances of /decl/hierarchy/skill have been created!</span>")
+			return 0
```
You shall never return anything in `/New()`, `CRASH()` if you must.

afterthought2 replied:
No longer returning. Still logging the error.

5crownik007 commented:
I'd have skillpoints scale with age but also make something get worse to balance it out
so we don't have 100 year old maint techs doing surgery
For now this looks gucci but once we get like attributes like strength and stuff make it so that age = more skillpoints but lower atts

afterthought2 commented:
Currently you don't accumulate any additional skill points past age 46, and the age modifier is not huge. This can be adjusted easily as well, including to decrease the skill buff for extremely old characters (though I think there is also a hard age limit in the code). It can also be modified for species that age differently.

I don't have any plans for implementing attributes myself. If others want to, of course they are free to try.

comma commented:
Pretty nice, got few nitpicks, mostly a TODO for myself to do after it's merged:
-colors kinda confusing, IMO should go with yellow for mandatory, green for taken, plain blue for possible
-thing is pretty hidden away, took me a bit to locate where do oyu get the skills
-maybe have higher levels of skill cost 2 points, currently pretty easy to end up with too much skill in jobs that don't have much specialization (e.g. explorer) but this might be adjusted with number of points they get too

afterthought2 commented:
@comma, the numbers are pretty arbitrary, and I'm sure they could benefit from some adjustment. The UI could probably use some improvement, but my design/artistic skills are nonexistent, so please, by all means. The colors are at least trivial to change in the file.

The balancing mechanism with skill selection is meant to be diminishing returns from higher investment in each skill encouraging generalization, but if this works poorly we could make different levels cost different amounts.

One thing to keep in mind, though: certain changes to the system will wipe saved data (if the saved skill allocation is no longer valid), so making big changes all the time will eventually start to annoy people.

Edit: changed colors, added a line to the job selection page about clicking on jobs to choose skills, squashed. New colors and description:
![skills4](https://user-images.githubusercontent.com/36152617/38527196-8824b244-3c27-11e8-81cc-c2efd6814877.png)

PsiOmegaDelta commented on code/modules/mob/mob.dm:
```diff
@@ -4,6 +4,8 @@
 	GLOB.living_mob_list_ -= src
 	unset_machine()
 	QDEL_NULL(hud_used)
+	if(istype(skillset))
```
`QDEL_NULL()` already makes its own appropriate checks.

afterthought2 replied:
This throws runtimes without the type check when the mob is deleted before initialization, as the thing there is then not a datum (but a type path).

PsiOmegaDelta commented on code/modules/mob/skills/skillset_silicon.dm:
```diff
@@ -0,0 +1,2 @@
+/datum/skillset/silicon
```
Could've just been 
```
/datum/skillset/silicon/obtain_from_mob(mob/mob)
    return
```
Removing a var and var check in the base `/datum/skillset/proc/obtain_from_mob(var/mob/mob)`

afterthought2 replied:
That would still allow skill transfer from silicon to carbon. You could check the type of skillset of mob in the base proc instead of the var, but I thought this would make it easier to make other things nontransferable.

PsiOmegaDelta commented on code/modules/mob/skills/skillset.dm:
```diff
@@ -0,0 +1,80 @@
+	. = ..()
+
+/datum/skillset/proc/from_datum(decl/hierarchy/skill/skill)
+	return ((skill in skill_list) ? skill_list[skill] : default_value)
```
Might as well just do ```return skill_list[skill] || default_value)```
BYOND handles getting non-existing keys, the original line above simply results in multiple enumerations of the associative list.

PsiOmegaDelta commented on code/modules/client/preferences.dm:
```diff
@@ -31,7 +31,15 @@ datum/preferences
+	var/job_max_skill	= list()
+
 /datum/preferences/New(client/C)
+	jobs_by_type = decls_repository.get_decls(GLOB.using_map.allowed_jobs)
```
Why are you even re-caching these values? The `decls_repository` already handles this.
Just call `decls_repository.get_decls(GLOB.using_map.allowed_jobs)` where appropriate.

PsiOmegaDelta commented on code/modules/client/preferences.dm:
```diff
@@ -31,7 +31,15 @@ datum/preferences
 
+	var/jobs_by_type
+	var/job_min_skill	= list()
+	var/job_max_skill	= list()
```
What exactly are the min and max skill lists for?
Do they somehow contain different values for different clients?

afterthought2 commented:
Conflicts fixed, again. I don't plan on making further changes to this PR unless one of the devs provides additional comments; it's been ready for review for some time.
